### ISSUE TITLE
[Feature] Support for Federated Repositories

### ISSUE THREAD

**Issue by feliu** [jfrog/terraform-provider-artifactory#92]
Is there any plans for adding support for [Federated Repositories](https://www.jfrog.com/confluence/display/JFROG/Federated+Repositories)?

It will be nice also to cover the use case of converting from a Local Repository to a Federated Repository. 
Context:https://www.jfrog.com/confluence/display/JFROG/Working+with+Federated+Repositories#WorkingwithFederatedRepositories-ConvertingaLocalRepositorytoaFederatedRepository

I'm happy to discuss the case and contribute to it. Thanks!

**Comment by feliu** [comment]
This is a first draft of the feature: https://github.com/jfrog/terraform-provider-artifactory/pull/93

Let me know what you think cc:@chb0github

**Comment by mmclane** [comment]
I am definitely going to need this!

**Comment by chb0github** [comment]
@feliu - So far, no complaints. But it seems like you're not ready - if you need an instance to test with I can supply one

**Comment by mmclane** [comment]
For now I have a working solution using the restapi resource and curl commands.  Its not early as pretty but it works, at least for creation.

**Comment by cawilson** [comment]
FYI - I've been working on the changes required in https://github.com/jfrog/jfrog-client-go to support this and have a PR open that's been going through review (see https://github.com/jfrog/jfrog-client-go/pull/375). Which also includes filling in gaps with repository types / parameters. Once this is through, my plan was to attempt changes in this provider to bring in support.

**Comment by cawilson** [comment]
Update: Support for Federated repositories have now been released for the `jfrog-client-go` library https://github.com/jfrog/jfrog-client-go/releases/tag/v1.3.0

**Comment by cawilson** [comment]
@chb0github - I've just started looking at adding Federated repositories to this provider and spending time enhancing https://github.com/jfrog/jfrog-client-go/ to find changes that appear to use, but also not use your own library. Are you able to provide more information on what the direction might be? Why would you not lean on `jfrog-client-go` but instead only use it for the structure but use Resty to make the API call, which is duplicating work?

Additionally, from what I can tell, the changes merged into `master` are broken for all resources except for the `artifactory_local_repository` resource which was modified. This was due to the changes in the `provider.go` file and removing the instantiation of the other libraries. This will put a major block on any additional changes until all other resources are changed or some of the code in `provider.go` is restored.

**Comment by chb0github** [comment]
Yes, all is true.

Right now, I am putting the finishing touches on v2.3.1 (on 2.3.x branch). The object of v2.3.1 is the full removal of the atlassian client code, with as much backward compatibility as possible. This necessitated some things. 

The artifactory go client had several incompatibilities with how terraform needs to function. In some cases, literal contradictions. The go client not only executes the calls for you, but also interprets the results and this more than a few times resulted in unresolvable problems. Some of the resources supported by this provider have no analog in the new client at all, for reasons I don't know. 

This wasn't always the case, but the overriding idea is to remove all the damn clients from this repo; 1 approach. Period. Sometimes implementing a resource using the go client and sometimes having to hack around it with resty worked contrary to this goal.

As for the master branch: I fall on my sword. I accidentally merged some of the code for these changes into the branch. However, I made the decision that, since 2.3.x was my goal, all would be fixed with that release. If you want to start working on federated repos, you can work off the 2.3.x branch.

Side note: It's my intention to release 2.3.1 in the next day-or-so, and merge in terraform SDK 2.0, if you can wait. 

I hope this all helps and sorry if it's a bit long

**Comment by cawilson** [comment]
@chb0github thanks for the detailed response. I appreciate the challenges with inheriting all this from Atlassian and trying to bring it forward and easier to manage. I was hoping that use the Jfrog Go client it would result in less duplication in code and where there were gaps, update the Go library and move forward. I do still have a branch in which I attempted to switch over the local repository to use the Go library and it did result in a number of switch statements but it was untested.

Maybe I will hold off for a moment and then revisit once the code settles down on the 2.3.1 version. I will have a look at the branch ahead of the merge.

I do have some ideas, which I wanted to explore, which includes separating the repository specific options into separate parameter groups and removing the flat structure. We could always deprecated the existing parameters, and when upgrading, consumers will just need to switch to use the new parameter but the underlying rest call won't change, so won't result in any changes, but will make it easier to understand which options apply to which repository type.

**Comment by chb0github** [comment]
2.3.x has settled down (and, for reasons, the current version is 2.6.8). I have begun to rollout repository specific resources now in order to take advantage of feature flags.

I have also discovered that the JF-go lib is lacking. But even putting that aside, some of the opinionated code is just incompatible with how TF needs to work. So, I have borrowed the structure where possible and stolen and modified where necessary. 

If you'd like to help contribute to federated repos, I am happy to talk

**Comment by bmanuel** [comment]
@alexhung, thank you for getting this support in. I am not seeing the documentation for these resources at https://registry.terraform.io/providers/jfrog/artifactory/latest/docs. Is there another process that needs to be completed before they are available for use?

**Comment by alexhung** [comment]
@bmanuel Ha! You're right! We forgot to add them.
